lib/events/recorder/recorder.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package recorder
+
+import (
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/events/filesessions"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// NewRecorder returns a [events.SessionRecorder]. If session recording is disabled,
+// a recorder is returned that will discard all session events. If session
+// recording is set to be synchronous, the returned recorder will use
+// syncStream to create an event stream. Otherwise, a streamer will be
+// used that will back recorded session events to disk for eventual upload.
+func NewRecorder(recCfg types.SessionRecordingConfig, cfg events.SessionWriterConfig, uploadDir string, syncStream events.Streamer) (events.SessionRecorder, error) {
```

Comment:
Hmm, this feels like a broken abstraction. There's 3 possible outcomes here in the happy path:

1. we produce a discard stream because recording is disabled.
2. we are doing async recording to disk
3. we are streaming in sync mode

We have to pass an upload dir, even though it's only used in 1 of 3 scenarios, and we have to pass a sync streamer even though it's only used in 1 of 3 scenarios.

At this point, what good is the `NewRecorder` helper func if you have to create everything ahead of time and pass it in?

Reply:
I see what you mean, it's not a perfect solution by any means, but all services that record sessions duplicated this exact logic before. My thought was to centralize it somewhere to prevent new services from configuring a session recorder incorrectly, as even with all these changes handling events it not super straightforward.